content/what-is-jsx.md

```diff
@@ -0,0 +1,255 @@
+<article class="blog-post">
+	<h2>Hello World</h2>
+	<p>lorem ipsum</p>
+</article>
+```
+
+becomes
+
+```javascript
+createElement("article", { class: "blog-post" },
+	createElement("h2", null, "Hello World"),
+	createElement("p", null, "lorem ipsum"));
+```
+
+And now we can leverage the power of JSX to not only create existing HTML
+elements but to write our own element definitions.
+
+For instance:
+
+```jsx
+<MyComponent title="Hello World">
+	<p>lorem ipsum</p>
+</MyComponent>
+```
+
+will be translated to
+
+```javascript
+createElement(MyComponent, { title: "Hello World" },
+	createElement("p", null, "lorem ipsum"));
+```
+
+Here JSX recognizes an element which is capitalized as being a user defined
+component and will pass `MyComponent` as a variable reference into the
+`createElement` function.
+
+This is the point in the documentation where the JSX specification and the
+JSX implementation usually become hopelessly intertwined.
+In the following image we can see where the distinction occurs.
+
+![A file with JSX blocks is transpiled into JavaScript with createElement calls. These createElement calls are then interpreted by the JSX implementation](jsx-transpilation-and-interpretation.svg)
+
+The JSX specification defines how a JavaScript transpiler should translate JSX blocks
+into a series of `createElement` calls. But the `createElement` function
+comes from a JSX _implementation_ and will determine how the actual components
+need to be defined. We will go into the specific details about how the complate
+implementation below, but first we want to look at a few more details about the
+JSX implementation.
+
+<!--
+	here I want HTML so that I can define an anchor element to the section and
+	link to it!
+-->
+
+### Interpolation with JavaScript Expressions
+
+Everything which appears in a `{ }` block will be interpreted as a JavaScript
+expression.
+
+```jsx
+<MyComponent title={title}>
+    <p>{description}</p>
+</MyComponent>
+```
+
+will become
+
+```javascript
+createElement(MyComponent, { title: title },
+	createElement("p", null, description));
+```
+
+This means that we can use JavaScript expressions to express conditionals
+
+```jsx
+<MyComponent title={title || "unknown"}>
+    {description && (
+        <p>{description}</p>
+    )}
+</MyComponent>
+```
+
+or for a loop
+
+```jsx
+<List>
+    {items.map(item => (
+        <span>{item}</span>
+    ))}
+</List>
+```
+
+### Special Cases for Boolean Attributes
+
+Attributes without an explicit value in a JSX expression are interpreted as
+having the boolean value `true`. This means that the following two JSX blocks
+are identical:
+
+```jsx
+<button disabled>
+```
+
+```jsx
+<button disabled={true}>
+```
+
+### Fragment provided by JSX Implementation
+
+JSX always expects exactly one root element in an expression, but in some cases
+we want to return several at once. To do this, a JSX implementation will
+provide a `Fragment` component with provides an artificial hierarchy level.
```

Comment:
Not necessarily. Babel (and other compilers, I presume) supports swapping out `Fragment` with something else. In the text, I would treat `<></>` as the user interface and `Fragment` as an implementation detail.

Reply:
Well, yeah, the same way you don't have to call the function `React.createElement` (or even just `createElement`) - but both spec and established convention refer to this as `Fragment`. Indeed, as of now, AFAIK most complate users avoid the `<>…</>` shorthand in favor of `<Fragment>…</Fragment>`, so it seems important not to insist on the shorthand?

Reply:
The (current?) spec doesn't know about `<Fragment>`, it's an implementation detail of React.

But disregarding the spec, there's a better argument for not using it. The `<>` syntax is – IMHO – not a shorthand, but just syntax. `<Fragment>` on the other hand refers to an actual entity. Syntax that looks like an entity reference but has special meaning is best avoided.

And your comparison with `React.createElement` doesn't work here because we never reference `React.createElement` in JSX user code. But we do refer to `Fragment` in JSX user code. The former having special semantics is fine, the latter not.

Reply:
I guess I was (intentionally) conflating `Fragment` and [`JSXFragment`](https://facebook.github.io/jsx/#syntax):
> ```
> PrimaryExpression :
>     JSXElement
>     JSXFragment
> ```

I understand your point, I think, but it seems ... excessively formalistic? As in, for users, even those interested in understanding JSX vs. React vs. complate, that strikes me as a distinction without a difference? It's possible my assessment here is influenced by JSX's evolution though (the shorthand syntax is fairly new IIRC; it used to be just `<Fragment>`).